### PR TITLE
feat: add `did:ebsi` provider Veramo plugin

### PR DESCRIPTION
**Pull request by plesastapevka**
Add Veramo plugin for `did:ebsi` DID method for legal entities. It enables users:
- creating and onboarding a DID (A bearer token is required for onboarding which can be claimed [here](https://app-pilot.ebsi.eu/users-onboarding/v2)),
- resolving a `did:ebsi` DID.

### CODE REVIEW

File: packages/snap/src/index.ts [hunk 1]
```diff
@@ -94,6 +94,8 @@ export const onRpcRequest: OnRpcRequestHandler = async ({
         return ResultObject.success(res);
       case 'switchDIDMethod':
         isValidSwitchMethodRequest(request.params);
```
Comment (andyv09) [inline packages/snap/src/index.ts:96]: Maybe update `isValidSwitchMethodRequest` to verify ebsiBearer is a 'valid' string.

File: packages/snap/src/utils/didUtils.ts [hunk 2]
```diff
@@ -24,12 +29,10 @@ export async function changeCurrentVCStore(
...
 
-export async function getCurrentDid(
-  ethereum: MetaMaskInpageProvider,
-  state: SSISnapState,
-  account: string
-): Promise<string> {
+export async function getCurrentDid(params: ApiParams): Promise<string> {
+  const { state, account } = params;
   const method = state.accountState[account].accountConfig.ssi.didMethod;
+  // SO FAR DO SEM VSE GOOD
```
Comment (andyv09) [inline packages/snap/src/utils/didUtils.ts:35]: remove

File: packages/snap/src/utils/ebsiUtils.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+import { ApiParams } from '../interfaces';
+import { getAgent } from '../veramo/setup';
+import { snapGetKeysFromAddress } from './keyPair';
+
+export async function getDidEbsiIdentifier(
+  params: ApiParams,
+  args: IDIDManagerCreateArgs
+): Promise<string> {
+  // args.options.bearer je bearer token
```
Comment (andyv09) [inline packages/snap/src/utils/ebsiUtils.ts:13]: remove

**Review by andyv09 — COMMENTED** [review]
Can you also update types and connector for `switchDidMethod` function and add some tests in snap

File: packages/snap/src/did/ebsi/ebsiDidOnboarding.ts [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+
+  const payload = {
+    id: `urn:did:${uuidv4()}`,
+    '@context': ['https://www.w3.org/2018/credentials/v1'],
+    type: ['VerifiablePresentation'],
+    holder: args.identifier.did,
+    verifiableCredential: [verifiableAuthorization],
+  } as EbsiVerifiablePresentation;
+
+  // FIXME: find a way around this function, it uses a library which prevents the
```
[the remark this answers is not in the document]
Comment (plesastapevka) [reply packages/snap/src/did/ebsi/ebsiDidOnboarding.ts:56]: not yet, tracked in #161.

**Comment by plesastapevka** [comment]
> Can you also update types and connector for `switchDidMethod` function and add some tests in snap

@andyv09 I added the request params validation. Tests will be implemented in the PR for #161.

File: packages/snap/src/utils/didUtils.ts [hunk 5]
```diff
@@ -38,28 +40,48 @@ export async function getCurrentDid(
...
+    const didUrl = getDidEbsiKeyIdentifier(state, account);
+    return `did:key:${didUrl}`;
+  }
+  if (method === 'did:ebsi') {
+    const didUrl = await getDidEbsiIdentifier(params, {
+      provider: method,
+      kms: 'web3',
+      options: { bearer: params.ebsiBearer },
+    });
+    console.log(didUrl);
```
Comment (martines3000) [inline packages/snap/src/utils/didUtils.ts:53]: remove

File: packages/types/src/params.ts [hunk 6]
```diff
@@ -94,6 +94,7 @@ export type ResolveDIDRequestParams = {
 
 export type SwitchMethodRequestParams = {
   didMethod: AvailableMethods;
+  ebsiBearer?: string;
```
Comment (martines3000) [inline packages/types/src/params.ts:97]: Why do we need this in the switch method ?

File: packages/snap/src/utils/ebsiUtils.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      options: {
+        ...args.options,
+        privateKey: res?.privateKey,
+        id: keccak256(Buffer.from(res?.privateKey as string)).slice(2, 18), // usually random 16 bytes, in our case first 16 bytes of keccak hashed priv key
+      },
+    });
+    return identifier.did;
+  } catch (e) {
+    console.log(e);
+    return '';
```
Comment (martines3000) [inline packages/snap/src/utils/ebsiUtils.ts:36]: Better to `throw new Error("Failed to create new EBSI Identifier˙  <- something like that

File: packages/types/src/params.ts (hunk 6 above)
Comment (plesastapevka) [reply packages/types/src/params.ts:97]: The bearer token must be passed to the `agent.didManagerCreate()` function when trying both to create and onboard the EBSI identifier. We can see the `ebsiBearer` being assigned and passed on [here](https://github.com/blockchain-lab-um/ssi-snap/blob/9ae4616a086edbaed4b56fbc5e9a696cebdd5298/packages/snap/src/index.ts#L98). The frontend which will be using our Snap for managing EBSI legal entities' DIDs will need to create an EBSI bearer token field, into which users will paste the token they have claimed at [the official EBSI onboarding site](https://app-pilot.ebsi.eu/users-onboarding/v2). Currently, they support claiming the bearer token via reCAPTCHA and EU Login; therefore, fetching it cannot be done programmatically. The most *"barebones"* user onboarding API call we can make is [this one](https://api-pilot.ebsi.eu/docs/apis/users-onboarding/latest#/operations/post-users-onboarding-api-sessions), which still needs the token from reCAPTCHA/EU Login as the parameter.

Comment (martines3000) [reply packages/types/src/params.ts:97]: image

Comment (martines3000) [reply packages/types/src/params.ts:97]: Better to show a popupa and promp the user to input it. The UI is already available to do this. 

And if I am not mistaken we do this only the first time for `did:ebsi` (Legal entity) right ?

Comment (plesastapevka) [reply packages/types/src/params.ts:97]: Okay, agreed, thanks. However, the whole integration will be implemented in #162.

File: packages/snap/src/did/ebsi/ebsiDidUtils.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+export function privateKeyJwkToHex(privateKeyJwk: jose.JWK): string {
+  if (!privateKeyJwk.d) {
+    throw new Error('Key does not contain private key material');
+  }
+  return Buffer.from(privateKeyJwk.d, 'base64').toString('hex');
+}
+
+export function isOwner({
```
Comment (martines3000) [inline packages/snap/src/did/ebsi/ebsiDidUtils.ts:52]: Do we need this ?

Comment (plesastapevka) [reply packages/snap/src/did/ebsi/ebsiDidUtils.ts:52]: nope, forgot to remove.

**Review by martines3000 — APPROVED** [review]
:partying_face: :tada: Great work!

**Pull request merged**